contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -113,6 +117,86 @@ convert_blob(struct agent_connection* con, const char *blob, DWORD blen, char **
 	return success? 0: -1;
 }
 
+/*
+ * in current_user sub tree under key_name key
+ * remove all sub keys with value name value_name_to_remove
+ * and value data value_data_to_remove
+ */
+static int
+remove_matching_subkeys_from_registry(HKEY user_root, wchar_t* key_name, wchar_t* value_name_to_remove, char *value_data_to_remove) {
+	int index = 0, success = 0;
+	DWORD data_len;
+	HKEY root = 0, sub = 0;
+	char *data = NULL;
+	wchar_t sub_name[MAX_KEY_LENGTH];
+	DWORD sub_name_len = MAX_KEY_LENGTH;
+
+	if (RegOpenKeyExW(user_root, key_name, 0, DELETE | KEY_ENUMERATE_SUB_KEYS | KEY_WOW64_64KEY, &root) != 0) {
+		success = 1;
+		goto done;
+	}
+
+	while (1) {
+		sub_name_len = MAX_KEY_LENGTH;
+		if (sub) {
+			RegCloseKey(sub);
+			sub = NULL;
+		}
+		if (RegEnumKeyExW(root, index++, sub_name, &sub_name_len, NULL, NULL, NULL, NULL) == 0) {
+			if (RegOpenKeyExW(root, sub_name, 0, KEY_QUERY_VALUE | KEY_WOW64_64KEY, &sub) == 0 &&
+				RegQueryValueExW(sub, value_name_to_remove, 0, NULL, NULL, &data_len) == 0) {
+
+				if (data)
+					free(data);
+				data = NULL;
+
+				if ((data = malloc(data_len + 1)) == NULL ||
+					RegQueryValueExW(sub, value_name_to_remove, 0, NULL, data, &data_len) != 0)
+					goto done;
+				data[data_len] = '\0';
+				if (strncmp(data, value_data_to_remove, data_len) == 0) {
+					if (RegDeleteTreeW(root, sub_name) != 0)
+						goto done;
+					--index;
+				}
+			}
+		}
+		else
+			break;
```

Comment:
Do we want to return success if `RegEnumKeyExW()` fails with a transient error?

Reply:
Good catch. In any error except `ERROR_NO_MORE_ITEMS` we probably don't want to return success because the removal of all potential matches is not guaranteed to have happened. I've made a change to reflect that.